src/tools/remote-test-client/src/main.rs

```diff
@@ -68,20 +83,33 @@ fn spawn_emulator(target: &str, server: &Path, tmpdir: &Path, rootfs: Option<Pat
         start_qemu_emulator(target, rootfs, server, tmpdir);
     }
 
+    let timeout = connect_timeout();
+    let mut timed_out = true;
+
     // Wait for the emulator to come online
-    loop {
+    let mut total_dur = Duration::from_secs(0);
+    while total_dur < timeout {
         let dur = Duration::from_millis(2000);
         if let Ok(mut client) = TcpStream::connect(&device_address) {
             t!(client.set_read_timeout(Some(dur)));
             t!(client.set_write_timeout(Some(dur)));
             if client.write_all(b"ping").is_ok() {
                 let mut b = [0; 4];
                 if client.read_exact(&mut b).is_ok() {
+                    timed_out = false;
                     break;
                 }
             }
         }
         thread::sleep(dur);
+        total_dur += dur;
+    }
+
+    if timed_out {
```

Comment:
hmm, feel wrong to loop with `timed_out = true` since we didn't time out yet, so let's change to `successful_read` and inverse all booleans, so here we'd have `!successful_read`.